Sources/GRPC/GRPCServerPipelineConfigurator.swift

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Logging
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+import NIOTLS
+
+/// Configures a server pipeline for gRPC with the appropriate handlers depending on the HTTP
+/// version used for transport.
+///
+/// If TLS is enabled then the handler listens for an 'TLSUserEvent.handshakeCompleted' event and
+/// configures the pipeline appropriately for the protocol negotiated via ALPN. If TLS is not
+/// configured then the HTTP version is determined by parsing the inbound byte stream.
+final class GRPCServerPipelineConfigurator: ChannelInboundHandler, RemovableChannelHandler {
+  internal typealias InboundIn = ByteBuffer
+  internal typealias InboundOut = ByteBuffer
+
+  /// The server configuration.
+  private let configuration: Server.Configuration
+
+  /// Reads which we're holding on to before the pipeline is configured.
+  private var bufferedReads = CircularBuffer<NIOAny>()
+
+  /// The current state.
+  private var state: State = .notConfigured
+
+  private enum State {
```

Comment:
This doesn't seem to know what it's _expecting_ to see out of ALPN or text. That means in principle if the timing gets weird you can be tricked by the ALPN arriving late. This is unlikely in a proper NIO pipeline, but it seems like it might be better to know which thing we're expecting?

Reply:
Yeah that's fair I think, we should be more explicit about this.